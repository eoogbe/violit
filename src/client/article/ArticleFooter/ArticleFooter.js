/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useRef, useState } from 'react';
import PropTypes from 'prop-types';
import { MdDelete, MdMoreVert } from 'react-icons/md';
import styled from 'styled-components';
import useOnClickOutside from 'use-onclickoutside';
import { DropdownMenu } from 'client/dropdown';

const StyledArticleFooter = styled.footer`
  display: flex;
  flex-direction: column;
  color: ${({ theme }) => theme.gray700};
`;

const OverflowDropdown = styled.nav`
  display: flex;
  align-items: center;
  align-self: flex-end;
  position: relative;
`;

const OverflowToggle = styled.button`
  padding: ${({ theme }) => theme.spacingSm}
  border: none;
  background: transparent;
`;

OverflowToggle.displayName = 'OverflowToggle';

/** The footer of a self-contained, user-generated composition. */
const ArticleFooter = ({ contentId, deleteContent }) => {
  const [overflowOpen, setOverflowOpen] = useState(false);

  const overflowEl = useRef();
  useOnClickOutside(overflowEl, () => setOverflowOpen(false));

  const overflowToggleIconAttrs = {
    title: 'More actions',
    viewBox: '0 0 24 24',
    focusable: false,
  };

  const overflowMenuId = `overflow-menu-${contentId}`;

  const overflowMenuItems = [
    {
      onClick: deleteContent,
      icon: MdDelete,
      text: 'Delete',
    },
  ];

  return (
    <StyledArticleFooter>
      <OverflowDropdown ref={overflowEl}>
        <OverflowToggle
          type="button"
          aria-haspopup="true"
          aria-expanded={overflowOpen}
          aria-controls={overflowMenuId}
          onClick={() => setOverflowOpen(!overflowOpen)}
        >
          <MdMoreVert size="1.5em" attr={overflowToggleIconAttrs} />
        </OverflowToggle>
        {overflowOpen &&
          <DropdownMenu
            id={overflowMenuId}
            items={overflowMenuItems}
            placement="auto"
          />}
      </OverflowDropdown>
    </StyledArticleFooter>
  );
};

ArticleFooter.propTypes = {
  /** The unique id for the content. */
  contentId: PropTypes.string.isRequired,
  /** Deletes the content. */
  deleteContent: PropTypes.func.isRequired,
};

export default ArticleFooter;
