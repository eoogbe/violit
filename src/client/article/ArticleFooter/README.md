```jsx
<ArticleFooter contentId="article1" deleteContent={() => {}} />
```
