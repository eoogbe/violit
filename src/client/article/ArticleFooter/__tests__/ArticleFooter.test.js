/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { mount } from 'enzyme';
import * as R from 'ramda';
import TestRenderer from 'react-test-renderer';
import ArticleFooter from '../ArticleFooter';

it('renders without crashing', () => {
  const renderer = TestRenderer.create(
    <ArticleFooter contentId="article1" deleteContent={() => {}} />
  );
  expect(renderer.toJSON()).toMatchSnapshot();
});

it('renders the open menu', () => {
  const renderer = TestRenderer.create(
    <ArticleFooter contentId="article1" deleteContent={() => {}} />
  );
  const instance = renderer.root;

  const toggle = instance.find(
    R.pathEq(['type', 'displayName'],
    'OverflowToggle')
  );
  TestRenderer.act(() => {
    toggle.props.onClick();
  });
  expect(renderer.toJSON()).toMatchSnapshot();
});

it('deletes the content', () => {
  const deleteContent = jest.fn();
  const wrapper = mount(
    <ArticleFooter contentId="article1" deleteContent={deleteContent} />
  );

  wrapper.find('OverflowToggle').simulate('click');
  wrapper.find('MenuItemButton').first().simulate('click');

  expect(deleteContent).toHaveBeenCalled();
});
