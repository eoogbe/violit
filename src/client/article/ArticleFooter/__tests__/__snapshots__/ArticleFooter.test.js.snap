// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the open menu 1`] = `
<footer
  className="sc-htpNat gPlOuz"
>
  <nav
    className="sc-bxivhb eysMmV"
  >
    <button
      aria-controls="overflow-menu-article1"
      aria-expanded={true}
      aria-haspopup="true"
      className="sc-ifAKCX cLgJZS"
      onClick={[Function]}
      type="button"
    >
      <svg
        fill="currentColor"
        focusable={false}
        height="1.5em"
        size="1.5em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        title="More actions"
        viewBox="0 0 24 24"
        width="1.5em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
        />
      </svg>
    </button>
    <ul
      className="sc-bdVaJa dxqlea"
      id="overflow-menu-article1"
      role="menu"
    >
      <li
        role="menuitem"
      >
        <button
          className="sc-bwzfXH hxEwoI"
          onClick={[Function]}
          type="button"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            height="1.25em"
            size="1.25em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1.25em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
          Delete
        </button>
      </li>
    </ul>
  </nav>
</footer>
`;

exports[`renders without crashing 1`] = `
<footer
  className="sc-htpNat gPlOuz"
>
  <nav
    className="sc-bxivhb eysMmV"
  >
    <button
      aria-controls="overflow-menu-article1"
      aria-expanded={false}
      aria-haspopup="true"
      className="sc-ifAKCX cLgJZS"
      onClick={[Function]}
      type="button"
    >
      <svg
        fill="currentColor"
        focusable={false}
        height="1.5em"
        size="1.5em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        title="More actions"
        viewBox="0 0 24 24"
        width="1.5em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
        />
      </svg>
    </button>
  </nav>
</footer>
`;
